### ISSUE TITLE
Rules for shadowing built-in functions are unclear

### ISSUE THREAD

**Issue by jimblandy** [gpuweb/gpuweb#2619]
In the section "Declaration and Scope", the WGSL spec says:

> When an identifier is used in scope of one or more declarations for that name, the identifier will denote the object of the non-[module-scope](https://gpuweb.github.io/gpuweb/wgsl/#module-scope) declaration appearing closest to that use, or the module-scope declaration if no other declaration is in scope.

However, "Built-in Functions" says:

> Since a built-in function is always in scope, it is an error to attempt to redefine one or to use the name of a built-in function as an [identifier](https://gpuweb.github.io/gpuweb/wgsl/#identifier) for any other kind of declaration.

This suggests that redefinition is an error *because* the built-in function is always in scope. But the "closest to use" rule describes how to choose among multiple definitions in scope, which implies that having overlapping scopes is not a problem. So the redefinition isn't an error for that particular reason.

Possible solutions:

1) Forbid redefining the names of built-in functions in any scope. This is what I think "Built-in Functions" was aiming for.

2) Allow redefinition of built-in function names at any scope. Since it would be natural to assume that built-in functions have a scope that ends at the end of the program, this would require a clarification of the "same end of scope" rule, so that the user's module-scope definitions could shadow the built-in functions.

3) Allow redefinition of built-in function names at any non-module scope. This is what I think "Declaration and Scope" was aiming for.

For what it's worth, both Tint and Naga accept the following program, even though `distance` is a built-in function:

```
fn f(x: f32) -> f32 {
   let distance = x;
   let factor: f32 = 2.0;
   return distance / factor;
}
```

**Comment by alan-baker** [comment]
My preference would be 3 until we allow users to overload functions.

**Comment by dneto0** [comment]
Whoops.  I believe the intent is 3.
I think this was an oversight when shadowing scopes were added to the spec.  Mea culpa. :-)

**Comment by alan-baker** [comment]
This was addressed by #2621. I'll it to @jimblandy to close if he's satisfied.

**Comment by jimblandy** [comment]
Yes, that answers my question.
